# October 2024 abalone WGS workshop

<img src="https://github.com/user-attachments/assets/eb4b6eeb-1e28-4c67-90b8-2d543725006d" alt="drawing" width="400"/>

## Agenda
Over the course of the workshop the primary aim is to become acquainted with `snpArcher` in order to call variants on your new data. In addition to that, we will cover some terminal/server basics, steps to make your workflow easier, and some downstream analyses, including those relying on **genotype likelihoods** instead of variant calls.

This workshop is intended to be flexible, so topics can be added or dropped as needed depending on interest and skill level. Just communicate with me and we'll make adjustments!

Last thing - if you have suggestions for what to add to these workshop pages for your future self or other researchers, please let me know. I want to make this as useful as possible.

## Main modules
- [Setup](https://github.com/twooldridge/workshop/blob/main/terminal.md): Most of this will be done for you prior to the workshop (thanks Miguel!), but feel free to reference this in the future if you find yourself working in a different space.
- [Command line familiarity, working on servers](https://github.com/twooldridge/workshop/blob/main/misc.md): The basics of working on the command line and on a server. Far from complete!
- [snpArcher setup and execution](https://github.com/twooldridge/workshop/blob/main/snpArcher.md): Configuring snpArcher and running on whole genome data
- [Evaluating snpArcher results](https://github.com/twooldridge/workshop/blob/main/results.md): Basic walkthrough of initial snpArcher output, as well as some instructive follow-up analyses in `R`/`python`/`bash`
- [ANGSD and genotype likelihoods](https://github.com/twooldridge/workshop/blob/main/angsd.md): PCA, diversity & selection stats, etc. all without variant calls
- [Advanced tasks](): GitHub, jupyter notebook workflow, etc.
